test/test_custom_op_testing.py

```diff
@@ -0,0 +1,327 @@
+            return x.view_as(x)
+
+        def foo_meta(x):
+            return x.view_as(x)
+
+        lib.impl("foo", Foo.apply, "Autograd")
+        lib.impl("foo", foo_impl, "CPU")
+        lib.impl("foo", foo_meta, "Meta")
+
+        def f(x):
+            x = x.clone()
+            y = op(x)
+            x.sin_()
+            return y
+
+        x = torch.tensor(3.14159 / 3, requires_grad=True)
+        with self.assertRaisesRegex(
+                RuntimeError,
+                'Argument x is not defined to alias output but was aliasing'):
+            operator_compile_check(f, (x,), {})
+
+    def test_missing_abstract_impl(self, device):
+        lib = self.lib()
+        lib.define("foo(Tensor x) -> Tensor")
+        op = self.ns().foo.default
+
+        class Foo(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                with torch._C._AutoDispatchBelowAutograd():
+                    return op(x)
+
+            @staticmethod
+            def backward(ctx, gx):
+                return 2 * gx
+
+        def foo_impl(x):
+            return torch.tensor(x.cpu().numpy() ** 2, device=x.device)
+
+        lib.impl("foo", Foo.apply, "Autograd")
+        lib.impl("foo", foo_impl, "CPU")
+        lib.impl("foo", foo_impl, "CUDA")
+
+        def f(x):
+            y = op(x)
+            return y.sum(0)
+
+        x = torch.tensor([0, 1.], requires_grad=True)
+        with self.assertRaisesRegex(
+                torch._subclasses.fake_tensor.UnsupportedOperatorException,
+                '_test_custom_op.foo.default'):
+            operator_compile_check(f, (x,), {})
+
+    def test_incorrect_abstract_impl(self, device):
+        lib = self.lib()
+        lib.define("foo(Tensor x) -> Tensor")
+        op = self.ns().foo.default
+
+        class Foo(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                # Emulate AutoDispatchBelowADInplaceOrView, which is not bound into python
+                guard = torch._C._AutoDispatchBelowAutograd()
+                guard2 = torch._C.ExcludeDispatchKeyGuard(torch._C.DispatchKeySet(torch._C.DispatchKey.ADInplaceOrView))
+                try:
+                    return op(x)
+                finally:
+                    del guard
+                    del guard2
+
+            @staticmethod
+            def backward(ctx, gx):
+                return gx
+
+        def foo_impl(x):
+            return x ** 2
+
+        def foo_meta(x):
+            return x.unsqueeze(1) ** 2
+
+        lib.impl("foo", Foo.apply, "Autograd")
+        lib.impl("foo", foo_impl, "CPU")
+        lib.impl("foo", foo_impl, "CUDA")
+        lib.impl("foo", foo_meta, "Meta")
+
+        def f(x):
+            y = op(x)
+            return y.sum(0)
+
+        x = torch.tensor([0, 1.], requires_grad=True)
+        with self.assertRaisesRegex(
+                RuntimeError,
+                'FooBackward returned an invalid gradient at index 0'):
```

Comment:
Just curious, why is this the error message for this failure?

Reply:
AOTAutograd runs Foo's forward and backward.
On the forward, it accepts a tensor of shape [2] and returns a tensor of shape [1, 2].
On the backward, it accepts a tensor of shape [1, 2] and returns a Tensor of shape [1, 2]. Autograd ends up yelling.

We should have a dedicated test to check the FakeTensor rule (it's not in this PR because I'm not sure how to actually create FakeTensors in dynamic mode). I'll add that in a follow up.